### ISSUE TITLE
relationship with do expression proposal

### ISSUE THREAD

**Issue by Jack-Works** [gmarcos87/switch-return-proposal#2]
it seems like can be done with do-expr proposal: (I didn't check the completion value of switch stmt so I don't know if it really works)

```js
const val = do {
    switch (x) {
        case 1: "1"; break;
        case 2: "2"; break;
   }
}
```

**Comment by ljharb** [comment]
This is the correct answer to any attempt to make statements usable in expression position, imo.

**Comment by Jack-Works** [comment]
So you'd like to remove switch support in do expression just like `for` statement in do expression?

**Comment by ljharb** [comment]
@Jack-Works no, i think it makes sense for it to work in a do expression - I'm saying that "do expressions" should be the only path forward for all current statements being usable in expression position.

**Comment by Jack-Works** [comment]
So what about this? https://github.com/tc39/proposal-throw-expressions

**Comment by ljharb** [comment]
The committee decided that was the one exception, but i personally don’t think that proposal should have advanced at all - i just didn’t choose to block it.

**Comment by hax** [comment]
Though `do` works, it actually have a worse readability than this proposal, due to weird `value; break` pattern vs explicit `return value`.

**Comment by ljharb** [comment]
I’m not sure why you’d ever need break writing idiomatic code inside a do expression.

**Comment by hax** [comment]
I don't understand. What "idiomatic code"?

**Comment by ljharb** [comment]
If you mean the term, from a quick google:
> Idiomatic code is code that does a common task in the common way for your language. 

As for which code, I’m referring to the code you referenced that would use a weird break pattern.

**Comment by hax** [comment]
I mean, as Jack's code show, if u want to use `do { switch ... }` you always need such weird break pattern.

I think both of us could agree such break pattern is weird and bad. So the question is, whether it's the issue of do expression or switch.

I guess you will say it's only the issue of `switch`, but I think do expression also have issue, because such weirdness is not only `switch`, but also `for` (even worse) and `finally`, and `if` without `else` (not very bad, but still...). The problem behind it is we losing the explicit return in do expression compare to current IIFE pattern.

**Comment by hax** [comment]
...continue

And this is why the author of this proposal want 

```js
const color = switch(type) {
  case 'apple':
    return 'red'
  case 'banana':
    return 'yellow'
  default
    return 'invisible'
}
```

it could also apply to `for` and `if` cases.

I don't mean to support this proposal, but I'm trying to understand the logic and the real requirements/preference behind this proposal.
